### PR TITLE
adding rwmix parameters to the schema

### LINKED ISSUE #44: Add rwmixread and rwmixwrite parameters
## Please describe what you would like to see in this project

Per subject.

### PR DESCRIPTION
**Pull request by dustinblack**
## Changes introduced with this PR

Adds rwmixread and rwmixwrite parameters. Fixes #44 

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by dbutenhof — APPROVED** [review]
Given that "the latter of the two will [...] override the first", it seems that having *both* is just plain silly. However, I suppose it allows you to specify the split using whichever side you prefer, and you're just propagating `fio`'s silliness, so, fine.

**Comment by dustinblack** [comment]
> Given that "the latter of the two will [...] override the first", it seems that having _both_ is just plain silly. However, I suppose it allows you to specify the split using whichever side you prefer, and you're just propagating `fio`'s silliness, so, fine.

I agree it's silly (of fio) to have both. But it's possible to have an input file with inheritance from a "global"

**Comment by webbnh** [comment]
> the latter of the two will be used to override the first

How do I, as an Arcaflow user (or a workflow writer) know which one is "the latter"?  There is no ordering in the interface at this level...right?

On the other hand, I suspect that we as Arcaflow developers can determine at the Arca-FIO interface which of the two will always override the other...and so we really should document _that_.

(And, it would be _great_ if we had a test which could confirm that that behavior is maintained (e.g., by checking the ordering of the switches on the command line).)

**Review by mfleader — APPROVED** [review]
If there are only two choices, then you can always compute the complement. I agree with @dbutenhof. It seems fine since we're just propagating upstream silliness.

File: arcaflow_plugin_fio/fio_schema.py [hunk 1]
```diff
@@ -297,6 +297,34 @@ class JobParams:
...
+        validation.min(1),
+        validation.max(100),
+        schema.name("Read/Write Mix - Write"),
+        schema.description(
+            "Percentage of a mixed workload that should be writes. If both rwmixread "
+            "and rwmixwrite is given and the values do not add up to 100%, the latter "
+            "of the two will be used to override the first. This may interfere with a "
+            "given rate setting, if fio is asked to limit reads or writes to a certain "
+            "rate. If that is the case, then the distribution may be skewed. Default: "
+            "50."
```
Comment (jaredoconnell) [inline arcaflow_plugin_fio/fio_schema.py]: Instead of "is given", it should be "are given", since "both" is a plural pronoun.

**Comment by jaredoconnell** [comment]
> > the latter of the two will be used to override the first
> 
> How do I, as an Arcaflow user (or a workflow writer) know which one is "the latter"? There is no ordering in the interface at this level...right?
> 
> On the other hand, I suspect that we as Arcaflow developers can determine at the Arca-FIO interface which of the two will always override the other...and so we really should document _that_.
> 
> (And, it would be _great_ if we had a test which could confirm that that behavior is maintained (e.g., by checking the ordering of the switches on the command line).)

Is the order based on the plugin-generated FIO input file? If so, that would then depend on how your code orders its representation of the config from the schema input. If that's the case, then the plugin defines the order, and we should document it based on the plugin's implementation instead of the generic documentation for FIO.

**Comment by dustinblack** [comment]
> > > the latter of the two will be used to override the first
> > 
> > 
> > How do I, as an Arcaflow user (or a workflow writer) know which one is "the latter"? There is no ordering in the interface at this level...right?
> > On the other hand, I suspect that we as Arcaflow developers can determine at the Arca-FIO interface which of the two will always override the other...and so we really should document _that_.
> > (And, it would be _great_ if we had a test which could confirm that that behavior is maintained (e.g., by checking the ordering of the switches on the command line).)
> 
> Is the order based on the plugin-generated FIO input file? If so, that would then depend on how your code orders its representation of the config from the schema input. If that's the case, then the plugin defines the order, and we should document it based on the plugin's implementation instead of the generic documentation for FIO.

Yeah, the ordering I see is an issue. The order precedence comes from the underlying tool, but it's non-deterministic for us. I'll think about how best to address this.

**Review by webbnh — APPROVED** [review]
Marking them as conflicting seems like a very good solution.

**Review by jaredoconnell — APPROVED** [review]

File: arcaflow_plugin_fio/fio_schema.py (hunk 1 above)
Comment (jaredoconnell) [reply arcaflow_plugin_fio/fio_schema.py]: Issue no longer applies due to rewrite of documentation.

**Review by dbutenhof — APPROVED** [review]

**Pull request merged**